src/services/bunker.py

```diff
@@ -1,15 +1,413 @@
+        """
+        per_epoch_buckets = defaultdict(list[Validator])
+        for validator in all_slashed_validators:
+            v = validator.validator
+            if not v.slashed:
+                raise Exception("Validator should be slashed to detect slashing epoch range")
+            possible_slashed_epoch = int(v.withdrawable_epoch) - EPOCHS_PER_SLASHINGS_VECTOR
+            if int(v.withdrawable_epoch) - int(v.exit_epoch) > MIN_VALIDATOR_WITHDRAWABILITY_DELAY:
+                per_epoch_buckets[possible_slashed_epoch].append(validator)
+                continue
+            else:
+                for epoch in range(possible_slashed_epoch, ref_epoch + 1):
+                    per_epoch_buckets[epoch].append(validator)
+
+        return per_epoch_buckets
 
-    def is_bunker_mode(self) -> bool:
-        pass
+    @staticmethod
+    def _get_per_epoch_lido_midterm_penalties(
+        per_epoch_buckets: dict[EpochNumber, list[Validator]],
+        lido_slashed_validators: list[Validator],
+        total_balance: Gwei,
+    ) -> dict[EpochNumber, dict[str, Gwei]]:
+        """
+        Iterate through per_epoch_buckets and calculate lido midterm penalties for each bucket
+        """
+        min_bucket_epoch = min(per_epoch_buckets.keys())
+        per_epoch_lido_midterm_penalties: dict[EpochNumber, dict[str, Gwei]] = defaultdict(dict)
+        for epoch, slashed_validators in per_epoch_buckets.items():
+            slashed_indexes = set(k.index for k in slashed_validators)
+            lido_validators_slashed_in_epoch = [v for v in lido_slashed_validators if v.index in slashed_indexes]
+            if not lido_validators_slashed_in_epoch:
+                continue
+            # We should calculate penalties according to bounded slashings in past EPOCHS_PER_SLASHINGS_VECTOR
+            min_bounded_epoch = max(min_bucket_epoch, EpochNumber(epoch - EPOCHS_PER_SLASHINGS_VECTOR))
+            bounded_slashed_validators = {}
+            for e, e_vals in per_epoch_buckets.items():
+                if min_bounded_epoch <= e <= epoch:
+                    for v in e_vals:
+                        if v.validator.pubkey not in bounded_slashed_validators:
+                            bounded_slashed_validators[v.index] = int(v.validator.effective_balance)
+            slashings = sum(bounded_slashed_validators.values())
+            adjusted_total_slashing_balance = min(
+                slashings * PROPORTIONAL_SLASHING_MULTIPLIER_BELLATRIX,
+                total_balance
+            )
+            for v in lido_validators_slashed_in_epoch:
+                effective_balance = int(v.validator.effective_balance)
+                penalty_numerator = effective_balance // EFFECTIVE_BALANCE_INCREMENT * adjusted_total_slashing_balance
+                penalty = penalty_numerator // total_balance * EFFECTIVE_BALANCE_INCREMENT
+                per_epoch_lido_midterm_penalties[epoch][v.index] = penalty
```

Comment:
`epoch` here is an expected slashed epoch, not a midterm epoch.
midterm epoch is calculated as: `validator.withdrawable_epoch - EPOCHS_PER_SLASHINGS_VECTOR // 2`

https://github.com/ethereum/consensus-specs/blob/dev/specs/bellatrix/beacon-chain.md#slashings

Reply:
Yes, you are right. Fixed